Add `dbg!` macro from `std`

niklasmohrin opened the pull request:
Hey there,

I really like the `dbg!` macro from the standard library and thought that it would make a good addition to the existing printing macros (and it was something that could be split off from #459 :D).

I am not sure about licensing, since the Rust standard lib is licensed under apache/mit and this file is licensed under gpl. Would that be a problem?

This is my first time contributing to the kernel, so if there is something wrong with commit mesasges or anything else, please point me in the right direction :)

ojeda commented:
> I am not sure about licensing, since the Rust standard lib is licensed under apache/mit and this file is licensed under gpl. Would that be a problem?

It is just a few lines, but if the code is mostly copy-pasted, then let's be on the safe side: please separate the macro into its own file so that you can use the original licence. Also please add a comment with who has the copyright / where the code is taken from.

> This is my first time contributing to the kernel, so if there is something wrong with commit mesasges or anything else, please point me in the right direction :)

Please take a look at `Documentation/process/submitting-patches.rst`.

ojeda commented on rust/kernel/print.rs:
```diff
@@ -439,3 +439,31 @@ macro_rules! pr_cont (
     )
 );
+
+/// Equivalent to [`std::dbg`](https://doc.rust-lang.org/std/macro.dbg.html), but uses `pr_debug`.
```
Please copy upstream's documentation (suitably changed).

ojeda commented on rust/kernel/print.rs:
```diff
@@ -439,3 +439,31 @@ macro_rules! pr_cont (
+/// Equivalent to [`std::dbg`](https://doc.rust-lang.org/std/macro.dbg.html), but uses `pr_debug`.
+#[macro_export]
+macro_rules! dbg {
+    // Comment from original / std implementation:
+
```
If we move to another file, and the top-level comment explains where the code comes from, then this can be removed.

ojeda commented on rust/kernel/print.rs:
```diff
@@ -439,3 +439,31 @@ macro_rules! pr_cont (
+    // `$val` expression could be a block (`{ .. }`), in which case the `eprintln!`
+    // will be malformed.
+    () => {
+        $crate::pr_debug!("[{}:{}]", core::file!(), core::line!());
```
Note that `pr_*` does not add a newline -- did you try it at runtime?

niklasmohrin replied:
It seems to not matter, the prints show up in their own line. Maybe its just dmesg being nice though, I can definetely add the newline for good practice.

I messed up a bit: While now testing this again (we used to have it with `pr_info` back at https://github.com/niklasmohrin/linux/blob/2d803fceab31195ff6fa8ad209b2fe0438f130cf/rust/kernel/print.rs), I saw that printing on the debug level is actually different from the other levels (I found https://stackoverflow.com/questions/28936199/why-is-pr-debug-of-the-linux-kernel-not-giving-any-output). When I copied it over from our branch and saw that there now is a `pr_debug`, I just replaced it with that thinking it would be more fitting. Now I find `pr_debug` to actually be rather unconvenient though and would prefer something like `pr_info` (should have checked again before opening, sorry for that).

If I (as a dev just wanting to print debug info) would have to set up the kernel with some debug flags or settings, I would be more inclined to just use `pr_info` with mediocre formatting instead of `dbg`. Maybe I am missing something though ... :)

I guess there could be a CI job that ensures that there are no `dbg!` calls in production code so that the `info` level log is not getting "spammed" if that would be a problem. What do you or others think?

ojeda replied:
> I guess there could be a CI job that ensures that there are no `dbg!` calls in production code so that the `info` level log is not getting "spammed" if that would be a problem. What do you or others think?

Sounds like a good idea -- could you please create an issue to about forgetting about it?

niklasmohrin replied:
:+1: see #511

niklasmohrin commented:
(no changes since last time, not done yet)

Thanks for your quick response :)

Should we use MIT or Apache? `std` is under the "mit or apache" license, but this is not a choice according at https://spdx.org/licenses/. Any reason not to use one or the other? From my quick googling, [GPL2 and apache v2 seem to slightly conflict (see paragraph starting with "despite our best efforts)"](https://www.apache.org/licenses/GPL-compatibility.html), whereas MIT seems to be okay ("The MIT license is compatible with many copyleft licenses, such as the GNU General Public License (GPL)", [wikipedia](https://en.wikipedia.org/wiki/MIT_License)).

Its also hard to point out a copyright holder other than maybe the Rust foundation? Or "the Rust developers"? Or there could just be a link to the github repository and https://rust-lang.org

> Please take a look at Documentation/process/submitting-patches.rst.

Will do :)

nbdd0121 commented:
> Should we use MIT or Apache? `std` is under the "mit or apache" license, but this is not a choice according at [spdx.org/licenses](https://spdx.org/licenses/). Any reason not to use one or the other? From my quick googling, [GPL2 and apache v2 seem to slightly conflict (see paragraph starting with "despite our best efforts)"](https://www.apache.org/licenses/GPL-compatibility.html), whereas MIT seems to be okay ("The MIT license is compatible with many copyleft licenses, such as the GNU General Public License (GPL)", [wikipedia](https://en.wikipedia.org/wiki/MIT_License)).

SPDX has [expressions](https://spdx.github.io/spdx-spec/appendix-IV-SPDX-license-expressions/#overview) and you can `OR` two licenses together.

When you only take code, it's okay to choose one license from multiple licenses; but this means you can no longer contribute back to the original project. Essentially, it's `MIT OR Apache-2.0` when taking the code, and `MIT AND Apache-2.0` when upstreaming the changes. So it's **essential** to keep the license as is.

Apache-2.0 is not compatible with GPL-2.0, this means that when the code is being compiled in the Linux kernel, we use the MIT license. But the Apache-2.0 still has to be retained.

EDIT: Also, Apache-2.0 is the *main* license for Rust. MIT is only there for GPL-2.0 compatibility.

ojeda commented:
> Should we use MIT or Apache? [...]
>
> Its also hard to point out a copyright holder other than maybe the Rust foundation? Or "the Rust developers"? Or there could just be a link to the github repository and https://rust-lang.org

Take a look at how I did it for our in-tree `alloc` in https://github.com/Rust-for-Linux/linux/tree/rust/rust/alloc.

niklasmohrin commented:
I think everything is worked in now

ojeda commented:
Looks good.

It seems I cannot re-run the PR actions before merging... Let's see.
